### PR TITLE
KMeans clustering and weighted centroid analysis

### PR DESCRIPTION
**Pull request by stuartlynn**
@ohasselblad CR please :-) 

Adds KMeans clustering and weighted centroid functions to crankshaft.

- [ ] All declared geometries are `geometry(Geometry, 4326)` for general geoms, or `geometry(Point, 4326)`
- [x] Include python is activated for new functions. Include this before importing modules: `plpy.execute('SELECT cdb_crankshaft._cdb_crankshaft_activate_py()')`
- [x] Docs for public-facing functions are written
- [x] New functions follow the naming conventions: `CDB_NameOfFunction`. Where internal functions begin with an underscore `_`.
- [x] If appropriate, new functions accepts an arbitrary query as an input (see [Crankshaft Issue #6](https://github.com/CartoDB/crankshaft/issues/6) for more information)

### CODE REVIEW

File: src/py/crankshaft/crankshaft/clustering/kmeans.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+from sklearn.cluster import KMeans
+import plpy
+
+def kmeans(query, no_clusters, no_init=20):
```
Comment (rochoa) [inline src/py/crankshaft/crankshaft/clustering/kmeans.py]: Why are we going with n_init=20? Default seems to be 10 at sklearn: http://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html.

I don't know too much about this but I guess we need to find the trade off between performance and accuracy. Do we know how that value affects both?

File: src/py/crankshaft/crankshaft/clustering/kmeans.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    data = plpy.execute('''select array_agg(cartodb_id order by cartodb_id) as ids,
```
Comment (rochoa) [inline src/py/crankshaft/crankshaft/clustering/kmeans.py]: As far as I understand this query we will only be able to use geo dimensions for kmeans, so we cannot use other type of dimensions for the distance calculation, right?

File: src/py/crankshaft/crankshaft/clustering/kmeans.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+from sklearn.cluster import KMeans
```
Comment (rochoa) [inline src/py/crankshaft/crankshaft/clustering/kmeans.py]: Should we declare this as a dependency?

File: src/py/crankshaft/crankshaft/clustering/kmeans.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import plpy
+
+def kmeans(query, no_clusters, no_init=20):
+    data = plpy.execute('''select array_agg(cartodb_id order by cartodb_id) as ids,
+        array_agg(ST_X(the_geom) order by cartodb_id) xs,
```
Comment (rochoa) [inline src/py/crankshaft/crankshaft/clustering/kmeans.py]: This will fail for null geometries as it will provide KMeans function with invalid data.

Comment (stuartlynn) [inline src/py/crankshaft/crankshaft/clustering/kmeans.py]: Yeah good catch. Will update to include a not null filter

File: src/py/crankshaft/crankshaft/clustering/kmeans.py (hunk 2 above)
Comment (stuartlynn) [inline src/py/crankshaft/crankshaft/clustering/kmeans.py]: Yeah right now its restricted to geo. No reason we couldn't use other variables as well but for this example we only needed to do it in the geo space. I think we should run with this now and perhaps write another more general function to do k-means over geo + variables. We will need to think about that a little more cause we need to assure that each variable is scaled appropriately

File: src/py/crankshaft/crankshaft/clustering/kmeans.py (hunk 1 above)
Comment (stuartlynn) [inline src/py/crankshaft/crankshaft/clustering/kmeans.py]: So K-Means is a stochastic algorithm which is not necessarily guaranteed to find the right answer with just 1 run. The n_init variable is the number of times the algorithm runs after with it takes the best answer, the one with the best defined clusters. 

The number n_init depends a lot on the problem, the data space and the number of classes. It's not 100% possible to predict it ahead of time. 

The performance should scale linearly with n_init. The accuracy is almost impossible to predict. 

Happy to lower it for now. Was just airing on the side of getting good clusters

File: src/py/crankshaft/crankshaft/clustering/kmeans.py (hunk 3 above)
Comment (stuartlynn) [inline src/py/crankshaft/crankshaft/clustering/kmeans.py]: yeah will add it to the setup.py file

File: src/py/crankshaft/crankshaft/clustering/kmeans.py (hunk 1 above)
Comment (rochoa) [inline src/py/crankshaft/crankshaft/clustering/kmeans.py]: Keep it like that, we could expose that later at camshaft level.

And thanks a lot for the explanation 😄 .

File: src/pg/sql/11_kmeans.sql [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    
+    import plpy 
+    plpy.execute('SELECT cdb_crankshaft._cdb_crankshaft_activate_py()')
+    from crankshaft.clustering import kmeans
+    return kmeans(query,no_clusters,no_init)
+
+$$ language plpythonu;
+
+CREATE OR REPLACE FUNCTION CDB_WeightedMean(query text, weight_column text, category_column text default null )
+RETURNS table (the_geom geometry,class integer ) as $$
```
Comment (rochoa) [inline src/pg/sql/11_kmeans.sql]: Why do we limit this to be an integer?

Example for this: I would like to have the neighbourhood name as the category. Does it make sense?

Wouldn't be better to cast to `::text` so we can support almost any column type? 

I don't know a better way because using polymorphic types would force us to have a more complex function signature, which I don't want.

File: src/pg/sql/11_kmeans.sql [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+BEGIN
+
+RETURN QUERY
+    EXECUTE format( $string$
+        select ST_SETSRID(st_makepoint(cx, cy),4326) the_geom, class  from (
+            select  
+                   %I as class,
+                   sum(st_x(the_geom)*%I)/sum(%I) cx,
```
Comment (rochoa) [inline src/pg/sql/11_kmeans.sql]: This will trigger a `division by zero` error when sum of `weight_column` is 0.

**Comment by stuartlynn** [comment]
Hey @rochoa I reworked CDB_WeightedMean in to an aggregate function 

```postgres
select  cdb_crankshaft.CDB_WeightedMean(the_geom, weight), class from table group by class
```
I think it makes a lot more sense and will resole some the other issues. It means tweaking camshaft though after you merged that PR. 

Let me know what you think

File: src/py/crankshaft/crankshaft/clustering/kmeans.py (hunk 4 above)
Comment (stuartlynn) [inline src/py/crankshaft/crankshaft/clustering/kmeans.py]: Fixed this here : https://github.com/CartoDB/crankshaft/pull/59/commits/9d3de5a8ef13be63539248f3e4d82d7b4b68df9d

**Comment by rochoa** [comment]
It's fine to upgrade camshaft as far as we don't change existing params :-). So feel free to open a new PR in camshaft changing the query using this new function signature. But: wasn't this suppose to run on the whole table/query? Sorry if I didn't understand this one.

**Comment by stuartlynn** [comment]
Aggregates run over the whole table but don't need the whole table all at once. The weighted sum can be broken down in to a running average over each row, so it can be formed as an aggregate but K-Means looks at the entire table at once so its harder to write as an aggregate. The KMeans function is still full table scanning.

Will open a new camshaft PR

File: src/py/crankshaft/crankshaft/clustering/kmeans.py (hunk 3 above)
Comment (stuartlynn) [inline src/py/crankshaft/crankshaft/clustering/kmeans.py]: Added here https://github.com/CartoDB/crankshaft/pull/59/commits/1a4944b9600250a972458bfe1952f79ffce76ff2

**Comment by rochoa** [comment]
Sorry, I totally missed the change from `CREATE FUNCTION` to `CREATE AGGREGATE`. Thanks for the explanation :-).

**Comment by rochoa** [comment]
Should we merge this one? 🎉

**Comment by stuartlynn** [comment]
Yup think it's good to go. Will let you do the honors @rochoa !
